_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,705 @@
+| actionpack     | ?                     |
+| activesupport  | ?                     |
+| railties       | ?                     |
+
+This left `actionpack`, `activesupport`, or `railties` as the location of `set_autoload_paths`. I needed to look at and search the source, but the rubygems pages for `actionpack`, `activesupport`, and `railties` had no "Source Code" link, so I went back to the rails gem page and clicked its ["Source Code" link](https://github.com/rails/rails).  Looking at the directory listing, I saw directories for `actionpack`, `activesupport`, and `railties`, so I assumed that all the gems were just housed in this same repository.  I needed to search against a specific tag of the repository and not how the code looks in master, so I cloned the repo to search locally
+
+```sh
+cd ~/git
+mkdir rails
+cd rails
+git clone git@github.com:rails/rails.git
+cd rails
+```
+
+I searched for `set_autoload_path` using `ack`, which is a nicer alternative to `grep`:
+
+```sh
+git checkout v3.2.2
+ack set_autoload_paths
+```
+
+I saw only 2 results
+
+![ack set_autoload_paths](/images/ack-set_autoload_paths.png)
+
+`railties/lib/rails/engine.rb` line 536 looks like a DSL call for setting up an initializer, so I opened `~/git/rails/rails/railties` in Rubymine to take advantage of its ability to jump from definition and usage easier than vim.  I did Shift+Shift and typed `engine.rb` and hit enter to open the file.  I did `:536` to jump to line `536` because I use IDEAVim to have vim bindings in Rubymine.
+
+![railties/lib/rails/engine.rb:536](/images/railties-lib-rails-engine-line-536.png)
+
+### `set_autoload_paths` implementation
+
+I see `ActiveSupport::Dependencies.autoload_paths` being added to with `unshift`, but I don't see `config.autoload_paths` being read except for the `config.autoload_paths.freeze` on line 541, so I Cmd+Click `_all_autoload_paths` to see it's definition.
+
+![all_autoload_paths](/images/all_autoload_paths.png)
+
+`_all_autoload_paths` is just the combination of `config.autoload_paths`, `config.eager_load_paths`, and `config.autoload_once_paths`, so if all I care about is emulating `set_autoload_paths` is I need to do `ActiveSupport::Dependencies.autoload_paths.unshift(config.autoload_paths)`, which is what I ended up doing in the PR.  Looking at the other lines in `set_autoload_paths`, I'll want to `freeze` my paths to prevent future modifications.
+
+### Emulating initializers without `Rails::Engine`
+
+So, that's take care of what my `set_autoload_paths` should do, but how am I going to get prosvc to do it without `Rails::Engine`?  I need to look more at how `Rails::Engine`s works to see which pieces I should copy to mimic its path handling from boot to set_autoload_paths.  I know that initializers are run in the order they are defined except where `:before` or `:after` would override that order from [the configuring guide](http://guides.rubyonrails.org/v3.2.2/configuring.html#rails-railtie-initializer).
```

Comment:
So, **~~that's take~~** **that takes** care of what my `set_autoload_paths` should do, but how am I going to get prosvc to do it without `Rails::Engine`?